pkg/repair/repaircontroller.go

```diff
@@ -0,0 +1,124 @@
+package repair
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"istio.io/pkg/log"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	client "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+
+	"k8s.io/client-go/util/workqueue"
+)
+
+type RepairController struct {
+	clientset     client.Interface
+	workQueue     workqueue.RateLimitingInterface
+	podController cache.Controller
+
+	reconciler BrokenPodReconciler
+}
+
+func NewRepairController(reconciler BrokenPodReconciler) (*RepairController, error) {
+	c := &RepairController{
+		clientset:  reconciler.client,
+		workQueue:  workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		reconciler: reconciler,
+	}
+
+	podListWatch := cache.NewFilteredListWatchFromClient(
+		c.clientset.CoreV1().RESTClient(),
+		"pods",
+		metav1.NamespaceAll,
+		func(options *metav1.ListOptions) {
+			labelSelectors := []string{}
+			fieldSelectors := []string{}
+
+			for _, ls := range []string{options.LabelSelector, reconciler.Filters.LabelSelectors} {
+				if ls != "" {
+					labelSelectors = append(labelSelectors, ls)
+				}
+			}
+			for _, fs := range []string{options.FieldSelector, reconciler.Filters.LabelSelectors} {
+				if fs != "" {
+					fieldSelectors = append(fieldSelectors, fs)
+				}
+			}
+			options.LabelSelector = strings.Join(labelSelectors, ",")
+			options.FieldSelector = strings.Join(fieldSelectors, ",")
+		},
+	)
+
+	_, c.podController = cache.NewInformer(podListWatch, &v1.Pod{}, 0, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(newObj interface{}) {
+			c.workQueue.AddRateLimited(newObj)
+		},
+		UpdateFunc: func(_, newObj interface{}) {
+			c.workQueue.AddRateLimited(newObj)
+		},
+	})
+
+	return c, nil
+}
+
+func (rc *RepairController) Run(threadiness int, stopCh <-chan struct{}) {
+	go rc.podController.Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh, rc.podController.HasSynced) {
+		runtime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+		return
+	}
+
+	for i := 0; i < threadiness; i++ {
```

Comment:
nit: seems unneeded since this is always 1, and we probably also always want it to be 1?